### ISSUE TITLE
Potential memory leak when deploy and undeploy repeatedly

### ISSUE THREAD

**Issue by alfred-qit** [espertechinc/esper#246]
Hi guys,

We are using esper 8.5.0 in our production environment. Recently we found that if we deploy and undeploy repeatedly, the memory usage of whole system kept increasing. 

here are the sample test codes:

`
import com.espertech.esper.common.client.EPCompiled;
import com.espertech.esper.common.client.configuration.Configuration;
import com.espertech.esper.common.client.module.Module;
import com.espertech.esper.common.client.module.ModuleItem;
import com.espertech.esper.compiler.client.CompilerArguments;
import com.espertech.esper.compiler.client.EPCompileException;
import com.espertech.esper.compiler.client.EPCompiler;
import com.espertech.esper.compiler.client.EPCompilerProvider;
import com.espertech.esper.runtime.client.*;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Test;

@Slf4j
public class RepeatedlyDeployTest {
    static final String RUNTIME_URI = "my-runtime-uri";

    @Test
    void testRepeatedlyDeploy() throws EPDeployException, EPUndeployException, EPCompileException {
        Configuration configuration = new Configuration();
        configuration.getCommon().addEventType(String.class.getSimpleName(), String.class.getName());

        EPRuntime runtime = EPRuntimeProvider.getRuntime(RUNTIME_URI, configuration);

        EPCompiled epCompiled = compile();

        System.gc();
        long startMemUsage = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        for (int i = 0; i < 2000; i++) {
            String deploymentId = deploy(runtime, epCompiled);
            undeploy(runtime, deploymentId);
        }
        System.gc();
        long endMemUsage = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
        log.info("start: {}, end: {}, substract: {}", startMemUsage, endMemUsage, endMemUsage - startMemUsage);
    }

    private String deploy(EPRuntime runtime, EPCompiled epCompiled) throws EPDeployException {

        EPDeployment deployment;
        deployment = runtime.getDeploymentService().deploy(epCompiled);

        return deployment.getDeploymentId();
    }

    private void undeploy(EPRuntime runtime, String deploymentId) throws EPUndeployException {
        runtime.getDeploymentService().undeploy(deploymentId);
    }

    private EPCompiled compile() throws EPCompileException {
        Configuration configuration = new Configuration();
        configuration.getCommon().addEventType(String.class.getSimpleName(), String.class.getName());
        CompilerArguments compilerArguments = new CompilerArguments(configuration);

        String normalCreateWindowEpl = "@public create window MyWindow#time(1 sec) String";

        Module module = new Module();
        module.getItems().add(new ModuleItem(normalCreateWindowEpl));

        EPCompiler compiler = EPCompilerProvider.getCompiler();

        return compiler.compile(module, compilerArguments);
    }
}
`

We suspect the deploy action will cause a great amount of runtime class been generated and won't be released since it's not garbage collectable.

We have tried changing to esper 8.7.0 but the problem remains. We are not sure if this is designed to be or a potential memory leak.

Can you guys reproduce this problem and maybe tell us the reason behind this?

**Comment by raguessner** [comment]
I need to check but I think "gc" only collects heap and classes do not get loaded into heap and get managed and released independently. You could try this out by setting a small heap and deploying and undeploying - it wouldn't run out of memory since heap memory does get freed by the runtime.

**Comment by gavol** [comment]
I have experienced something similar. In my case the leak was in the "Metaspace" memory pool and linked to the deploy and un-deploy operations.

I have observed thread local objects (several of type ByteArrayProvidingClassLoader) causing memory to not be released. I have solved the problem making sure that all the (un)deployment operations are always executed in a dedicated thread (most likely thread local objects are then re-used).

**Comment by alfred-qit** [comment]
> I have experienced something similar. In my case the leak was in the "Metaspace" memory pool and linked to the deploy and un-deploy operations.
> 
> I have observed thread local objects (several of type ByteArrayProvidingClassLoader) causing memory to not be released. I have solved the problem making sure that all the (un)deployment operations are always executed in a dedicated thread (most likely thread local objects are then re-used).

Yes this is what I think so far too. But still I don't know why classes are generated and loaded as "different class" even if I create named windows with the same name.

**Comment by bernhardttom** [comment]
Named windows that are public are unique by name. Named windows that are not public are unique by the combination of deployment id and name.
